### DISCUSSION TITLE
SSH to remote host not working

### DISCUSSION THREAD

**Discussion by Anz14** [StackStorm/st2#5636] [Operations] [answered]
I have stackstorm configured in Ubuntu VM. And I'm integrating stackstorm with alert-manager for auto remediation. When the alert triggers it matches the rule and trigger the action given. So my action is trying to login to the remote host which is triggering the alert and perform some actions. However SSH to remote host is not working. What needs to be done here to get the SSH working.

 I'm getting the below error while executing the action.

stderr': '05/02/2022 08:37:45 AM Executing the script\n05/02/2022 08:37:45 AM Trying to ssh on host **<IP ADDRESS**>\nTraceback (most recent call last):\n  File "/opt/stackstorm/st2/lib/python3.8/site-packages/python_runner/python_action_wrapper.py", line 395, in <module>\n    obj.run()\n  File "/opt/stackstorm/st2/lib/python3.8/site-packages/python_runner/python_action_wrapper.py", line 214, in run\n    output = action.run(**self._parameters)\n  File "/opt/stackstorm/packs/demo/actions/kafka_demo_disk_action.py", line 44, in run\n    status, output = ssh_connect(host)\n  File "/opt/stackstorm/packs/demo/actions/kafka_demo_disk_action.py", line 62, in ssh_connect\n    k = paramiko.RSAKey.from_private_key_file("/home/ubuntu/.ssh/stackstorm-admin-kafka.pem")\n  File "/opt/stackstorm/st2/lib/python3.8/site-packages/paramiko/pkey.py", line 235, in from_private_key_file\n    key = cls(filename=filename, password=password)\n  File "/opt/stackstorm/st2/lib/python3.8/site-packages/paramiko/rsakey.py", line 55, in __init__\n    self._from_private_key_file(filename, password)\n  File "/opt/stackstorm/st2/lib/python3.8/site-packages/paramiko/rsakey.py", line 175, in _from_private_key_file\n    data = self._read_private_key_file("RSA", filename, password)\n  File "/opt/stackstorm/st2/lib/python3.8/site-packages/paramiko/pkey.py", line 307, in _read_private_key_file\n    with open(filename, "r") as f:\nFileNotFoundError: [Errno 2] No such file or directory: \'/home/ubuntu/.ssh/stackstorm-admin-kafka.pem\'\n', 'exit_code': 1, 'result': 'None'}

**Comment by arm4b** [comment, marked as answer]
The error message you should pay attention to is:
```
FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/.ssh/stackstorm-admin-kafka.pem'
```

Looks like the SSH key file you're trying to use doesn't exist?

**Comment by Anz14** [comment]
@armab how can I configure my stackstorm to SSH to remote host?

**Reply by arm4b** [reply]
See https://docs.stackstorm.com/troubleshooting/ssh.html

**Comment by Anz14** [comment]
@armab I have done the troubleshooting as per the doc.

```
root@gw-uat-stackstorm:/home/stanley/.ssh# st2 run core.remote cmd=whoami hosts=test-zookerper-1 username=stanley private_key=/home/stanley/.ssh/stanley_rsa
.
id: 626fe49a76815794f3ab0289
action.ref: core.remote
context.user: st2admin
parameters:
  cmd: whoami
  hosts: test-zookerper-1
  private_key: '********'
  username: stanley
status: succeeded
start_timestamp: Mon, 02 May 2022 14:03:06 UTC
end_timestamp: Mon, 02 May 2022 14:03:08 UTC
result:
  test-zookerper-1:
    failed: false
    return_code: 0
    stderr: ''
    stdout: stanley
    succeeded: true
```


```
root@gw-uat-stackstorm:/home/stanley/.ssh# st2 run core.remote cmd=whoami hosts=test-zookerper-1
.
id: 626fe41376815794f3ab0280
action.ref: core.remote
context.user: st2admin
parameters:
  cmd: whoami
  hosts: test-zookerper-1
status: succeeded
start_timestamp: Mon, 02 May 2022 14:00:51 UTC
end_timestamp: Mon, 02 May 2022 14:00:52 UTC
result:
  test-zookerper-1:
    failed: false
    return_code: 0
    stderr: ''
    stdout: stanley
    succeeded: true
```

But still I'm getting below error :

`"stderr": "05/02/2022 01:56:45 PM Executing the script\n05/02/2022 01:56:45 PM Trying to ssh on host <IP ADDRESS>\nTraceback (most recent call last):\n  File \"/opt/stackstorm/st2/lib/python3.8/site-packages/python_runner/python_action_wrapper.py\", line 395, in <module>\n    obj.run()\n  File \"/opt/stackstorm/st2/lib/python3.8/site-packages/python_runner/python_action_wrapper.py\", line 214, in run\n    output = action.run(**self._parameters)\n  File \"/opt/stackstorm/packs/demo/actions/kafka_demo_disk_action.py\", line 44, in run\n    status, output = ssh_connect(host)\n  File \"/opt/stackstorm/packs/demo/actions/kafka_demo_disk_action.py\", line 62, in ssh_connect\n    k = paramiko.RSAKey.from_private_key_file(\"/home/ubuntu/.ssh/stackstorm-admin-kafka.pem\")\n  File \"/opt/stackstorm/st2/lib/python3.8/site-packages/paramiko/pkey.py\", line 235, in from_private_key_file\n    key = cls(filename=filename, password=password)\n  File \"/opt/stackstorm/st2/lib/python3.8/site-packages/paramiko/rsakey.py\", line 55, in __init__\n    self._from_private_key_file(filename, password)\n  File \"/opt/stackstorm/st2/lib/python3.8/site-packages/paramiko/rsakey.py\", line 175, in _from_private_key_file\n    data = self._read_private_key_file(\"RSA\", filename, password)\n  File \"/opt/stackstorm/st2/lib/python3.8/site-packages/paramiko/pkey.py\", line 307, in _read_private_key_file\n    with open(filename, \"r\") as f:\nFileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/.ssh/stackstorm-admin-kafka.pem'\n",`

In which server it's looking for  '/home/ubuntu/.ssh/stackstorm-admin-kafka.pem'

**Reply by arm4b** [reply]
Per your log, looks like the custom SSH key `/home/ubuntu/.ssh/stackstorm-admin-kafka.pem` is referenced by the custom demo action you're running `/opt/stackstorm/packs/demo/actions/kafka_demo_disk_action.py`

**Comment by Anz14** [comment]
In the action there was a reference to the SSH key in that particular location. I have changed that, its sorted now. Thanks for the help @armab
